Docs: svelte tutorial

jonniebigodes opened the pull request:
This PR introduces a first draft to the "Intro to Storybook" tutorial for svelte.

A live version is [here](https://jonniebigodes-storybook-svelte.netlify.com/) and the code used for writting the tutorial is [here](https://github.com/jonniebigodes/intro-storybook-svelte)

A few caveats with the tutorial:
- Composite component section
   - The decorator code and action was ommited. Reason being that despite my attempts to create a similar approach, using slots and also `<svelte:component>` to allow it to work in a similar fashion led to the component not actually loading at all and some other issues. Could not pinpoint to a actual cause, but this section can be later introduced with a future update to both packages.
- Screen section
   - The section was shortened a bit, as the case that is demonstrated with the other versions does not  apply. The way stores work with svelte are a bit different. Tried a couple of different approaches aswell, with [derived stores](https://svelte.dev/docs#derived) and even [xstate](https://xstate.js.org/). I'm doing some research on how to implement Redux with svelte and see if we can use it so that the document aligns itself with the remaining versions.
- Testing section
   - the chromatic api call is now ` npx chromatic --app-code=<app-code>` as reported by chromaticqa site. (The change could be added aswell to the other versions with a correction pr).
- Deploy section
   - A bit of work was done there, to split up the initialization of the repo locally and the github one
- Conclusion section
   - The link to the code will default to the master branch as no svelte tutorial is present. 
- Others:
  - npm was used throughout the documentation to keep it consistent.
  - The framework and language selector renders a bit weird as it should as there's no icon for this.

Feel free to provide feedback.

jonniebigodes commented:
@tmeasday i know that the team as been hard at work with a new release, but i was wondering if you managed to take a look at the documentation in this pr?

tmeasday commented:
Hey @jonniebigodes sorry, mate, I had started but let me try and get through it today

tmeasday commented on content/intro-to-storybook/svelte/en/screen.md:
```diff
@@ -0,0 +1,161 @@
+  import InboxScreen from "./components/InboxScreen.svelte";
+</script>
+
+<InboxScreen error={$AppStore} />
```
After doing this I found the app rendered correctly but I couldn't actually change the state of tasks.

tmeasday replied:
https://github.com/tmeasday/learnstorybook-svelte/commit/491504148e1def613c8ffc55ea4215fcf30a3194

jonniebigodes replied:
The data and screen code needs a bit of a fix. I've tracked down the error and i'll update the code tomorrow and commit it if you don't mind.

jonniebigodes replied:
The store code was updated to correctly display the data and allow the state changes with the latest commit

tmeasday commented on content/intro-to-storybook/svelte/en/test.md:
````diff
@@ -0,0 +1,154 @@
+npm install -D storybook-chromatic
+```
+
+Import Chromatic in your `.storybook/config.js` file.
````
This step isn't needed any more. We should probably remove it from all tutorials (cc @ndelangen)

ndelangen replied:
Indeed, this should be removed from all tutorials.

jonniebigodes replied:
It's now removed with the latest commit. Going to mark it as resolved if you both don't mind

tmeasday commented on content/intro-to-storybook/svelte/en/test.md:
```diff
@@ -0,0 +1,154 @@
+  />
+</video>
+
+Run the test command in the command line to setup visual regression tests for Storybook. Don't forget to add your unique app code in place of `<app-code>`.
```
We need to add the `-s` option to the `build-storybook` script here (or maybe we should do it at the start when we do it to the `storybook` script?

jonniebigodes replied:
My example code is using the following script for building storybook:
```json
{
 "scripts":{
   .....
"build-storybook": "build-storybook -c .storybook -s public -o storybook-static",
}
}
Is that ok with you?

tmeasday replied:
Sure, although I think the `-c` is not necessary (that's the default) and the `-o` doesn't necessarily matter for this tutorial. So it might be better to keep things simple?

jonniebigodes replied:
I've added some changes to the document to address this. See if the wording is ok with you and let me know.

tmeasday commented on content/intro-to-storybook/svelte/en/using-addons.md:
````diff
@@ -0,0 +1,235 @@
+First, we will need to install all the necessary dependencies.
+
+```bash
+npm install @storybook/addon-knobs
````
This should be `-D`

jonniebigodes replied:
It's now updated with the latest commit. All of the other tutorials should be also updated to reflect this changes, as they are instructing to add the addon as a dependency. Going to mark it as resolved if you don't mind

tmeasday commented on content/intro-to-storybook/svelte/en/creating-addons.md:
```diff
@@ -0,0 +1,473 @@
+---
```
I'm guessing this file is unchanged as it is not in Svelte (of course). LMK if not.

jonniebigodes replied:
The creating addons was moved from the existing one for the react version and was not changed.  If you don't mind i'm going to mark this as resolved.

jonniebigodes commented on content/intro-to-storybook/svelte/en/get-started.md:
```diff
@@ -0,0 +1,192 @@
+
+## Add assets
+
+We also need to add the font and icon [directories](https://github.com/chromaui/learnstorybook-code/tree/master/public) to the `public/` folder.
```
This one was one of those that got me wondering. But i went with the current flow of documentation. My initial thought that could be applied here and then moved into the other tutorials is to apply the same "strategy" used for the design systems tutorial, more specifically in [the architecting systems section](https://www.learnstorybook.com/design-systems-for-developers/react/en/architecture/) under "Create an inventory" . What do you think?

jonniebigodes replied:
I've updated it to match the design systems wording and structure. One thing i would like for you to check is the urls that i've used in conjunction with svn. I don't have svn installed at this moment so i was unable to test them. I changed them because the ones you supplied were throwing a 401 error with github.

tmeasday replied:
I think the URLs only work over svn:

<img width="799" alt="image" src="https://user-images.githubusercontent.com/132554/69112077-8db44400-0ad3-11ea-9f35-a8499d35331a.png">

jonniebigodes replied:
Updated the links with the latest commit.

jonniebigodes commented on content/intro-to-storybook/svelte/en/get-started.md:
````diff
@@ -0,0 +1,192 @@
+And remove the following tag:
+
+```html
+<link rel='stylesheet' href='/global.css'>
````
I think that copying the css into the `global.css` file and mentioning it in the documentation seems a bit more reasonable, avoids adding a extra dependency and messing with the config. I left it originally like this, but it seems that it's a better way that way and will make the get started more leaner. Thoughts?

tmeasday replied:
Yeah I think the simplest thing is best. We aren't aiming to teach people how to use CSS with Svelte.

jonniebigodes replied:
It's now simplified with the latest commit. Going to mark it as resolved if you don't mind.

jonniebigodes commented on content/intro-to-storybook/svelte/en/composite-component.md:
```diff
@@ -0,0 +1,256 @@
+import { withPinnedTasks } from "./TaskList.stories";
+describe("TaskList", () => {
+  it("renders pinned tasks at the start of the list", () => {
+    setTimeout(() => {
```
I was getting into some errors, when the tests ran. More specifically getting undefined errors, so that's the reason of the timeout.  Going to move the code for a async/await pattern for better readability if you don't mind

tmeasday replied:
I think I managed to make it work on my branch, take a look at https://github.com/tmeasday/learnstorybook-svelte/commit/a197f3dd8cf73fecc541782cf668a0e49fda4b55

jonniebigodes commented on content/intro-to-storybook/svelte/en/get-started.md:
```diff
@@ -0,0 +1,192 @@
+cd taskbox
+
+# Add Storybook:
+npx -p @storybook/cli sb init
```
In order for the documentation to move forward, if you don't mind i'm going to update the command to match the official instructions and we can later apply a small pr to fix this. Sounds good?

tmeasday replied:
That sounds great

tmeasday commented:
That sounds great
On 19 Nov 2019, 11:00 AM +1100, jonniebigodes <notifications@github.com>, wrote:
> @jonniebigodes commented on this pull request.
> In content/intro-to-storybook/svelte/en/get-started.md:
> > +Storybook runs alongside your app in development mode. It helps you build UI components isolated from the business logic and context of your app. This edition of Learn Storybook is for Svelte; other editions exist for [Vue](/vue/en/get-started), [Angular](/angular/en/get-started) and [React](/angular/en/get-started).
>
> +
>
> +![Storybook and your app](/intro-to-storybook/storybook-relationship.jpg)
>
> +
>
> +## Setup Svelte Storybook
>
> +
>
> +We’ll need to follow a few steps to get the build process set up in your environment. To start with, we want to use [Degit](https://github.com/Rich-Harris/degit) to setup our build system, and enable [Storybook](https://storybook.js.org/) and [Jest](https://facebook.github.io/jest/) testing in our created app. Let’s run the following commands:
>
> +
>
> +```bash
>
> +# Create our application:
>
> +npx degit sveltejs/template taskbox
>
> +
>
> +cd taskbox
>
> +
>
> +# Add Storybook:
>
> +npx -p @storybook/cli sb init
>
> In order for the documentation to move forward, if you don't mind i'm going to update the command to match the official instructions and we can later apply a small pr to fix this. Sounds good?
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

jonniebigodes commented on content/intro-to-storybook/svelte/en/get-started.md:
```diff
@@ -0,0 +1,192 @@
+# Create our application:
+npx degit sveltejs/template taskbox
+
+cd taskbox
```
I forgot about that. Svelte/degit doesn't install dependencies out of the box. Updated with latest commit.

jonniebigodes commented on content/intro-to-storybook/svelte/en/simple-component.md:
```diff
@@ -0,0 +1,341 @@
+            id:task.id
+        });
+    }
+    //
```
This was lapse from my end. I tend to use it as a personal coding style. When i added the code i forgot to remove them. Fixed with the latest commit. Going to mark it as resolved if you don't mind.

jonniebigodes commented on content/intro-to-storybook/svelte/en/simple-component.md:
````diff
@@ -0,0 +1,341 @@
+```javascript
+// .storybook/config.js
+
+import { configure } from '@storybook/react';
````
Alll fixed with the latest commit. Forgot to remove change it when i was writting and reading the tutorial. Fixed with latest commit. Going to mark it as resolved if you don't mind

jonniebigodes commented on content/intro-to-storybook/svelte/en/get-started.md:
````diff
@@ -0,0 +1,192 @@
+
+- The first one called `fileMock.js` with the following content:
+  ```javascript
+  module.exports='file-stub';
````
Think that it should be ok now. It should be formatted correctly. Can you check on your end?

jonniebigodes replied:
Think that it should be ok now. It should be formatted correctly. Can you check on your end?

tmeasday replied:
LGTM

jonniebigodes commented on content/intro-to-storybook/svelte/en/get-started.md:
```diff
@@ -0,0 +1,192 @@
+	....
+	plugins: [
+		....
+		css({output: 'public/index.css'}), //bundles the copied css into the app 
```
This should now be somewhat irrelevant as the css is copied into the `global.css` file. As per comment below

jonniebigodes commented on content/intro-to-storybook/svelte/en/simple-component.md:
```diff
@@ -0,0 +1,341 @@
+  onArchiveTask: action("onArchiveTask")
+};
+
+storiesOf("Task", module)
```
Going to change the stories format throughout the tutorial to match CSF tomorrow in conjunction with the data/screen sections.

tmeasday replied:
:+1: Do you want to it in this PR or separately?

jonniebigodes replied:
as this pr is still under review and not merged, i think it's the best time to tackle this and make the change at this stage and then move the approach to the other frameworks with a subsequent pr.

jonniebigodes commented on content/intro-to-storybook/svelte/en/composite-component.md:
```diff
@@ -0,0 +1,256 @@
+          tasks: withPinnedTasks
+        }
+      });
+      expect(container.firstChild.classList.contains("TASK_PINNED")).toBe(true);
```
Moved the tests to async/await pattern and now no errors are thrown. Let me know if you're ok with the change.

tmeasday replied:
Looks great

jonniebigodes commented:
> Looks great!
> 
> Did you have code commits somewhere? If not, I followed along and pushed my code here, with a few changes as commented:
> 
> https://github.com/tmeasday/learnstorybook-svelte

I left the url for the code i'm using in the pr description, in case you've missed it, it's [here](https://github.com/jonniebigodes/intro-storybook-svelte). Pushed a commit to address a portion of the feedback provided. I'll update the rest tomorrow, namely the data/screen section store parts and commit it then. Sounds good?

jonniebigodes commented on content/intro-to-storybook/svelte/en/screen.md:
```diff
@@ -0,0 +1,161 @@
+  };
+};
+
+export const AppStore = appState();
```
code was updated with latest commit. Marking this as resolved if you don't mind

jonniebigodes commented:
@tmeasday @domyen and @WagnerMoreira i've updated the deployment section to incorporate the wording mentioned in #222 see if it's ok and let me know if changes are required.

WagnerMoreira commented on content/intro-to-storybook/svelte/en/deploy.md:
```diff
@@ -0,0 +1,109 @@
+
+![Netlify connect to repo](/intro-to-storybook/netlify-account-picker.png)
+
+Configure Netlify by highlighting which build command to run in its CI and which directory the static site is outputted in. For branch choose `master`. Directory is `storybook-static`. Build command use `yarn build-storybook`.
```
Hi @jonniebigodes  please change it for Directory is `.storybook-static` as per PR https://github.com/chromaui/learnstorybook.com/pull/222

jonniebigodes replied:
will do it shortly. Forgot about that.

jonniebigodes replied:
Should be fixed now. Marking this as resolved if you don't mind

WagnerMoreira replied:
sounds good!

jonniebigodes commented:
@domyen and @tmeasday with the latest commits i've updated this version of the tutorial to csf format as discussed above. 
A few caveats that need to be addressed.
- In LSB users were allowed to name a story as `default`, as this is a reserved word i modified it to `standard` and with that the images/videos will probably need some tweaks to reflect the pattern. 
- Moved the data used (namely the task object and the tasks array) to a helper file called `storybook-helper.js`, reasoning behind this is that if the user exports the withPinnedTasks` inside a story and starts storybook he'll be greeted with a error. Also as the data used in the tutorial is used in various places, sounds reasonable that it should "live" under one place and accessed as needed. If you have any other ideas regarding this, let me know and i'll adjust accordingly.
- The deploy section was updated, more specifically as the script to build storybook as a static app is modified earlier in the test section, no need to add duplication of information and add confusion, also the directory is now referencing the correct folder name, as per #222. 
- Simple section was updated to use `babel-plugin-require-context-hook`, this because when using the `babel-plugin-macros` storybook was not picking up the stories with csf format.
- If you want to, feel free to grab the code from the repo i've used and mentioned above in the pr description and add a branch for svelte, so that when the reader gets to that option he can see a svelte implementation.

Feel free to provide feedback so that we can merge this, or continue to work on it until is ready to be merged.

tmeasday commented:
@jonniebigodes sorry things have been a little crazy for me for the last couple of days. I want to take a proper look at this, hopefully I'll be able to tomorrow.

jonniebigodes commented:
@tmeasday no need to appologize, I would rather if you take care of what you need to take care of and probably take a better look at it during the weekend if you have the time. In the meantime i'll continue to work on adapting the react documentation to csf to begin a standartization of the documentation as you mentioned above in one of the comments in the review. Sounds good?

tmeasday commented:
@jonniebigodes I'd like to have a think about some of the points you raised in https://github.com/chromaui/learnstorybook.com/pull/212#issuecomment-555771771 and figure out a good path forward before applying CSF to other frameworks.

Specifically I am wondering how best to pass the data around. What's our latest thinking here @shilman? We should make sure we set a good pattern in LSB.

jonniebigodes commented:
@tmeasday ok.I'll stash the updates, put the migration on the "back burner"(pardon the bad pun for now) and wait for a consensus to be reached. In the meantime if you don't mind i'll work on implementing the tutorial for react native to close #96 and continue to improve the documentation. Sounds good?

tmeasday commented:
Sounds great @jonniebigodes

tmeasday commented:
In theory this looks fine but I want to talk to @shilman about what we want to encourage in terms of passing data around.

shilman commented:
@jonniebigodes This is tremendous -- really excellent work!

My recommendation for how to deal with data in CSF:

```js
export default {
  title: ...
  excludeStories = /.*Tasks/;
  // OR excludeStories = ['defaultTasks', 'pinnedTasks', ...]
}

// export data for reuse
export const defaultTasks = [ ... ]
export const pinnedTasks = [ ... ]
...

// stories are also named exports
export const _default = () => (...)
export const pinned = () => (....)
...
```

tmeasday commented:
@shilman I want to chat to you about this, but I was thinking in LSF we could set `excludeStories` to something like `/.*Data/` in the `.storybook/config.js` and then have a naming convention like `defaultTasksData`. WDYT?

shilman commented:
@tmeasday WFM, and happy to chat whenever you'd like

tmeasday commented:
@jonniebigodes I chatted to @shilman today and I think we are in agreement on [my comment above](https://github.com/chromaui/learnstorybook.com/pull/212#issuecomment-558856752).

Would you like me to make a PR to the react/english guide to demonstrate what I mean?

jonniebigodes commented:
@tmeasday sure i would like to see how it can be done in the react tutorial and i'll go from there. Also reading what @shilman said in his comment it "clicked" that you could use default if you put a " _ " before it. Completely forgot about it. When i used the approach i was aiming for simplicity and have the data all stored in one place as it's being used in several places throughout the tutorial, but if we have a way to maintain the tutorial almost identical and don't introduce a breaking change it's fine by me. I'll be waiting for the pr and make the necessary adjustments to this one. In the meantime i'll continue to work on the react native tutorial.

tmeasday commented:
I think we may as well use a different word that not a reserved word even thought the `_` trick is a good one.

tmeasday commented:
> When i used the approach i was aiming for simplicity and have the data all stored in one place as it's being used in several places throughout the tutorial,

The thing about this is that in a larger project you end up "building up" the data across stories as the components get more complicated. If you have a single source of story data, you end up making it contain the "maximal" data (i.e. what you need for the most complicated component) which means you "over-fetch" for the simpler components, which can lead to bugs.

There are ways around that, of course, but this compositional approach works really well IMO and I want to advocate it in the tutorial :)

jonniebigodes commented:
it's fine by me and i fully understand the intention, what i forgot to mention on my previous comment was that as the tutorial is not that big, keeping the way that i've mentioned was not a bad idea, but like you said and you're absolutely correct, is when this is applied to larger projects, it leads to more problems than solutions.....
Regarding this

> I think we may as well use a different word that not a reserved word even thought the `_` trick is a good one.

I went with "standard" as it's a similar word, but like i mentioned, choosing another word leads to making the videos and images once again, going with `_` we might "get away with it" for now as the word is similar to the one used presently and when the reader sees the media content he/she might overlook it. Up to you on how to proceed.

tmeasday commented:
OK, here is a first pass @jonniebigodes: https://github.com/chromaui/learnstorybook-code/pull/21

tmeasday commented:
@jonniebigodes is this ready to merge assuming we'll resolve CSF separately?

jonniebigodes commented:
@tmeasday i have the csf portion almost done. I've seen your example and i'm going to update the rest of the examples and commit this tonight and tomorrow you can take a look at it and if nothing is wrong merge it. Sounds good?

jonniebigodes commented:
@tmeasday just finished updating the code examples for csf per feedback, i think i didn't miss anything, but i would like for you to take a look at it and see if all is good and we can merge this.

tmeasday commented on content/intro-to-storybook/svelte/en/simple-component.md:
````diff
@@ -92,46 +71,61 @@ import { actions, task } from './storybook-helper';
 });
 ```
 
+Due to some language restraints the default action was named `Default` as the lowercase it's a reserved word, the `excludeStories` prop was added to avoid Storybook treating both the `taskData` and `actionsData` as actual stories, you can read more about it [here](https://storybook.js.org/docs/formats/component-story-format/#non-story-exports).
````
> Due to some language restraints the default action was named `Default`

We are moving (SB-wide) to using uppercase names for story exports, so I don't think we need to make excuses for it, I would say let's just call it `Default` and not really say why, that's just what we call stories.

jonniebigodes replied:
@tmeasday see my latest commit. Simplified it a bit.

tmeasday replied:
LGTM

tmeasday commented:
You might want to bring some small text changes over from the PR I just made but I am OK to merge this when you are ready @jonniebigodes

jonniebigodes commented:
With the latest commit i removed some lingering references to the helper file originally created.

jonniebigodes commented:
@domyen thank you very much for the speedy icon implementation. It's added and it looks real nice. @tmeasday i think this one is done, we can merge this when you're ready.

tmeasday commented:
awesome!
